### PR TITLE
Feat (Project): Project and Release dependency network management

### PR DESCRIPTION
**Pull request by hoangnt2**
Issue: 
#1538 

### How To Test?
Refer #1538 

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

### CODE REVIEW

**Comment by tienlee** [comment]
Hello,

I would like to explain for this PR.
Based on Qiu's idea #1538, we made a first PR, you can check the source code to understand about the idea. This PR is independence and don't effect to other features. We do that because there are many related features if we apply the idea about Dependency Network feature to sw360. Therefore, we made eclipse:1538-dependency-network for this PR, but not eclipse:master.

When Dependency Network feature is done, we will merge it into the eclipse:master branch.

In this PR, please help me to review the solutions, and give me your comments.

In addition, @shi9qiu has listed the following features which will be affected：
• License Clearing tag
• Clearing Request
• ECC tag
• Rest API function
• Import/Export function
• Change Log function
• SPDX import

With your experience about sw360 system, please let me know if you find any other related functions. And support me during implement this feature.

Thank you very much.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java [hunk 1]
```diff
@@ -1632,4 +1635,276 @@ public ProjectData searchByTag(String tag, User user) {
...
+     * Trace is list index of each node.
+     */
+    public List<ReleaseLink> getReleaseLinksOfProjectNetWorkByTrace(List<String> trace, String projectId, User user){
+        Project project = repository.get(projectId);
+        String releaseNetwork = project.getReleaseRelationNetwork();
+        ObjectMapper mapper = new ObjectMapper();
+        List<ReleaseLinkJSON> listReleaseLinkJson;
+        List<ReleaseLink> linkedReleases = new ArrayList<>();
+        try {
+            listReleaseLinkJson = mapper.readValue(releaseNetwork, new TypeReference<List<ReleaseLinkJSON>>() {
```
Comment (tsdv-phongnv) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:1652]: Modify <List<ReleaseLinkJSON>> to <>

IDE note: Explicit type arguments <List<ReleaseLinkJSON>> can be replaced with <>

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java [hunk 2]
```diff
@@ -1632,4 +1635,276 @@ public ProjectData searchByTag(String tag, User user) {
...
         return repository.searchByType(type, user);
     }
+
+
+
+    /*
+     * This function will get list ReleaseLink from release network of project with trace from root to clicked release on screen
+     * Trace is list index of each node.
+     */
+    public List<ReleaseLink> getReleaseLinksOfProjectNetWorkByTrace(List<String> trace, String projectId, User user){
```
Comment (tsdv-phongnv) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:1645]: trace => traces

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java [hunk 3]
```diff
@@ -1632,4 +1635,276 @@ public ProjectData searchByTag(String tag, User user) {
...
+        Project project = repository.get(projectId);
+        String releaseNetwork = project.getReleaseRelationNetwork();
+        ObjectMapper mapper = new ObjectMapper();
+        List<ReleaseLinkJSON> listReleaseLinkJson;
+        List<ReleaseLink> linkedReleases = new ArrayList<>();
+        try {
+            listReleaseLinkJson = mapper.readValue(releaseNetwork, new TypeReference<List<ReleaseLinkJSON>>() {
+            });
+            ReleaseLinkJSON previousNode = listReleaseLinkJson.get(Integer.parseInt(trace.get(0)));
+            for (int i=1; i < trace.size(); i++){
```
Comment (tsdv-phongnv) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:1655]: Suggestions: 

ReleaseLinkJSON finalPreviousNode = previousNode;
traces.forEach(trace -> finalPreviousNode.getReleaseLink().get(Integer.parseInt(trace)));

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java [hunk 4]
```diff
@@ -1632,4 +1635,276 @@ public ProjectData searchByTag(String tag, User user) {
...
+
+        if (linkedProjects != null && !linkedProjects.isEmpty()) {
+            flattenNetworkForLinkedProject(linkedProjects, projectOrigin, releaseOrigin, clearingStatusList,
+                    user, isInaccessibleLinkMasked);
+        }
+
+        return clearingStatusList;
+    }
+
+    private void flattenNetworksForReleases(List<ReleaseLinkJSON>  listReleaseLinkJson,
```
Comment (tsdv-phongnv) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:1737]: Remove method not use

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java [hunk 5]
```diff
@@ -1632,4 +1635,276 @@ public ProjectData searchByTag(String tag, User user) {
...
+        linkedProjects.entrySet().stream().forEach(lp -> wrapTException(() -> {
+            String projId = lp.getKey();
+            String relation = ThriftEnumUtils.enumToString(lp.getValue().getProjectRelationship());
+            if (projectOrigin.containsKey(projId))
+                return;
+            Project linkedProjectById = getProjectById(projId, user);
+            projectOrigin.put(projId, SW360Utils.printName(linkedProjectById));
+            Map<String, String> row = createProjectCSRow(relation, linkedProjectById, clearingStatusList);
+            Map<String, ProjectProjectRelationship> subprojects = linkedProjectById.getLinkedProjects();
+            Map<String, ProjectReleaseRelationship> linkedReleases = linkedProjectById.getReleaseIdToUsage();
```
Comment (tsdv-phongnv) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:1812]: Remove unused variables

File: backend/src/src-components/src/main/java/org/eclipse/sw360/components/ComponentHandler.java [hunk 6]
```diff
@@ -643,4 +643,18 @@ public Map<PaginationData, List<Component>> getRecentComponentsSummaryWithPagina
...
+    public List<ReleaseLink> getReleasesAndOtherWithSameComponent(Map<String, ProjectReleaseRelationship> relations, User user) throws TException {
+        assertNotNull(relations);
+
+        List<ReleaseLink> releaseLinks = handler.getLinkedReleasesWithAccessibility(relations, user);
+        for (ReleaseLink releaseLink : releaseLinks){
+            Release release = getReleaseById(releaseLink.getId(),user);
+            List<Release> releasesWithSameComponent = getReleasesByComponentId(release.getComponentId(),user);
+            releaseLink.setReleaseWithSameComponent(releasesWithSameComponent);
+        }
+        System.out.println(releaseLinks.toString());
```
Comment (tsdv-phongnv) [inline backend/src/src-components/src/main/java/org/eclipse/sw360/components/ComponentHandler.java:657]: Remove System.out.println

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java [hunk 7]
```diff
@@ -1632,4 +1635,276 @@ public ProjectData searchByTag(String tag, User user) {
...
+            if (subprojects != null && !subprojects.isEmpty()) {
+                flattenNetworkForLinkedProject(subprojects, projectOrigin, releaseOrigin, clearingStatusList,
+                        user, isInaccessibleLinkMasked);
+            }
+
+            projectOrigin.remove(projId);
+            row.put("projectOrigin", String.join(" -> ", projectOrigin.values()));
+        }));
+    }
+    private Optional<ProjectLink> createProjectNetworkLink(String id, ProjectProjectRelationship projectProjectRelationship, String parentNodeId,
```
Comment (tsdv-phongnv) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:1834]: Check indent parameter. Need to check other methods too

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java [hunk 8]
```diff
@@ -1632,4 +1635,276 @@ public ProjectData searchByTag(String tag, User user) {
...
+                    List<ReleaseLinkJSON> listReleaseLinkJson;
+                    List<ReleaseLink> linkedReleases = new ArrayList<>();
+                    try {
+                        listReleaseLinkJson = mapper.readValue(releaseNetwork, new TypeReference<List<ReleaseLinkJSON>>() {
+                        });
+                        linkedReleases = convertFromReleaseLinkJSONToReleaseLink(listReleaseLinkJson, id, user, "", 0);
+                    } catch (JsonProcessingException e) {
+                        log.error("JsonProcessingException: " + e);
+                    } catch (SW360Exception e) {
+                        throw new RuntimeException(e.getMessage());
```
Comment (tsdv-phongnv) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:1860]: convertFromReleaseLinkJSONToReleaseLink  throws TException
 SW360Exception extends TException 

=> just catchTException

File: backend/src/src-projects/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java [hunk 9]
```diff
@@ -435,4 +436,22 @@ public int getMyAccessibleProjectCounts(User user) throws TException {
     public void sendExportSpreadsheetSuccessMail(String url, String recepient) throws TException {
         handler.sendExportSpreadsheetSuccessMail(url, recepient);
     }
+
+    @Override
+    public List<ReleaseLink> getReleaseLinksOfProjectNetWorkByTrace(String projectId, List<String> trace, User user) throws TException {
```
Comment (tsdv-phongnv) [inline backend/src/src-projects/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java:441]: Remove throws TException.

TException is never throws in method

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortalConstants.java [hunk 10]
```diff
@@ -657,6 +657,45 @@ public class PortalConstants {
     public static final boolean SSO_LOGIN_ENABLED;
     public static final boolean IS_COMPONENT_VISIBILITY_RESTRICTION_ENABLED;
 
+    public static final String RELEASES_IN_NETWORK = "releasesInNetwork";
+    public static final String FIND_SUB_LINKED_RELEASE = "findSubLinkedRelease";
+    public static final String LIST_SUB_LINKED_RELEASES = "listSubLinkedReleases";
```
Comment (tsdv-phongnv) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortalConstants.java:662]: Remove unnecessary unuse variable

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortalConstants.java [hunk 11]
```diff
@@ -657,6 +657,45 @@ public class PortalConstants {
...
+
+    public static final String INDEXES = "indexes[]";
+
+    public static final String PARENT_NODE_ID = "parentNodeIds[]";
+
+    public static final String COMMENTS = "comments[]";
+
+    public static final String DEFAULT_VALUES = "defaultValues[]";
+
+    public static final String RELEASE_ID_ARRAY = "releaseIdArray[]";
```
Comment (tsdv-phongnv) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortalConstants.java:682]: public static final String RELEASE_ID_ARRAY = "releaseIdArray[]";

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/LinkedReleasesAndProjectsAwarePortlet.java [hunk 12]
```diff
@@ -300,4 +305,119 @@ public Map<String, String> fillMainLineState(Set<String> releaseIds, ComponentSe
         }));
         return relMainLineState;
     }
+
+    protected  List<ReleaseLinkJSON> getNetworkLinkedRelease(List<Release> releases, User user){
```
Comment (tsdv-phongnv) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/LinkedReleasesAndProjectsAwarePortlet.java:309]: Remove 2 space -> 1 space (after protected)

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java [hunk 13]
```diff
@@ -3230,4 +3273,136 @@ private Set<String> getProjectGroups(List<Project> projects) {
...
+            }
+        } catch (TException e) {
+            log.error("Error getting releases!", e);
+            throw new PortletException("cannot get releases " + Arrays.toString(linkedIds), e);
+        }
+        request.setAttribute(RELEASES_IN_NETWORK, linkedReleases);
+        include("/html/utils/ajax/linkedReleaseInNetwork.jsp", request, response, PortletRequest.RESOURCE_PHASE);
+    }
+
+    protected void putLinkedReleasesNetworkWithAccessibilityInRequest(PortletRequest request, Project project, User user){
```
Comment (tsdv-phongnv) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:3357]: project, user unuse

Is it necessary to add this method?

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java [hunk 14]
```diff
@@ -3230,4 +3273,136 @@ private Set<String> getProjectGroups(List<Project> projects) {
...
+        User user = UserCacheHolder.getUserFromRequest(request);
+        String projectId = request.getParameter(DOCUMENT_ID);
+        List<Map<String, String>> clearingStatusList = new ArrayList<Map<String, String>>();
+        try {
+            clearingStatusList = client.getDependencyNetworkForListView(projectId, user);
+        } catch (TException e) {
+            log.error("Problem getting flat view of Clearing Status", e);
+        }
+        JSONArray clearingStatusData = createJSONArray();
+        for (int i = 0; i < clearingStatusList.size(); i++) {
```
Comment (tsdv-phongnv) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:3372]: use foreach not index

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/projects/includes/projects/dependencyNetworkDetail.jsp [hunk 15]
```diff
@@ -0,0 +1,650 @@
...
+                for(let i=0;i<licArr.length;i++){
+                    let licLink=$("<a></a>").attr("href", makeLicenseViewUrl(licArr[i])).text(licArr[i]);
+                    licLinkArr.push(licLink[0].outerHTML);
+                }
+                return licLinkArr.join(", ");
+            }
+            return "";
+        }
+
+        function renderActions(id, type, row)
```
Comment (tsdv-phongnv) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/projects/includes/projects/dependencyNetworkDetail.jsp:405]: use early return for this method

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/projects/includes/projects/dependencyNetworkDetail.jsp [hunk 16]
```diff
@@ -0,0 +1,650 @@
...
+                    let currentIndex = $(this).attr("data-index");
+                    let trace = [];
+                    trace.unshift(currentIndex);
+                    if (releaseLayer > 0) {
+                        for (let layer = releaseLayer - 1; layer >= 0; layer--) {
+                            trace.unshift($(this).prevAll('tr[data-layer="'+layer+'"][data-release-id="'+parentId+'"]').first().attr('data-index'));
+                            parentId = $(this).prevAll('tr[data-layer="'+layer+'"][data-release-id="'+parentId+'"]').first().attr('data-parent-release');
+                        }
+                    }
+                    console.log(trace);
```
Comment (tsdv-phongnv) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/projects/includes/projects/dependencyNetworkDetail.jsp:567]: Remove console.log

**Comment by tienlee** [comment]
This PR is to demonstrate Qiu's idea of this feature.
If you have no comments, please close or reject this PR.
We are creating a new PR based on the latest code (main branch) for this feature and will publish it soon.

**Comment by ag4ums** [comment]
Closing this... as there will be new PR for this

**Pull request closed without merging**